### ISSUE TITLE
Error in first run after upgrade to v3.0.0

### ISSUE THREAD

**Issue by sahmadreza** [aagarwal1012/Animated-Text-Kit#168]
Hi,
After updating the plugin version from 2.5.4 to **3.0.0**, an error will be displayed below the first run.

But after a few seconds the animation starts and works

I did not change the code and just updated the plugin and after the update this problem occurred

My Code:
```
                    SizedBox(
                      height: 100,
                      child: FadeAnimatedTextKit(
                        onTap: () {
                          print("Tap Event");
                        },
                        isRepeatingAnimation: true,
                        text: [
                          "do IT!",
                          "do it RIGHT!!",
                          "do it RIGHT NOW!!!"
                        ],
                        textStyle: TextStyle(
                            fontSize: 27.0,
                            fontWeight: FontWeight.w400,
                          color: Colors.white
                        ),
                        textAlign: TextAlign.start,
                      ),
                    ),

```
**Expected behavior**
```
════════ Exception caught by widgets library ═══════════════════════════════════════════════════════
The following NoSuchMethodError was thrown building AnimatedBuilder(animation: AnimationController#f1944(▶ 0.942), dirty, state: _AnimatedState#da5af):
The getter 'value' was called on null.
Receiver: null
Tried calling: value

The relevant error-causing widget was: 
  FadeAnimatedTextKit file:///E:/Projects/Android/robo_investment/lib/app.dart:112:30
When the exception was thrown, this was the stack: 
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      FadeAnimatedText.animatedBuilder (package:animated_text_kit/src/fade.dart:45:24)
#2      _AnimatedTextKitState.build.<anonymous closure> (package:animated_text_kit/src/animated_text.dart:187:32)
#3      AnimatedBuilder.build (package:flutter/src/widgets/transitions.dart:1443:19)
#4      _AnimatedState.build (package:flutter/src/widgets/transitions.dart:181:48)
...
════════════════════════════════════════════════════════════════════════════════════════════════════
```
**(package:animated_text_kit/src/fade.dart:45:24) :**
```
  @override
  Widget animatedBuilder(BuildContext context, Widget child) {
    return Opacity(
      opacity: **_fadeIn.value** != 1.0 ? _fadeIn.value : _fadeOut.value,
      child: textWidget(text),
    );
  }
```
**Screenshots**
https://i.imgur.com/2s1nTEo.png

**Flutter:**
 Flutter (Channel stable, 1.22.4, on Microsoft Windows [Version 10.0.18363.1082], locale en-US)
 - version 1.22.4
 - channel stable

**Dart:**
 - version  2.10.4

**Android Studio.**
 - version: 4.1.0

**Comment by kw2019ltd** [comment]
same issue

**Comment by awhitford** [comment]
I'm sorry to see this.  Honestly, this should be _impossible_ because `animatedBuilder` is called _after_ `initAnimation` which initializes `_fadeIn`.

**I am unable to reproduce this problem.**  Are you able to reproduce this with the `example` application?  It looks like you are using Windows for an Android target?  What Android device target are you using?

@kw2019ltd Are you using Windows for an Android target too?  What Android device target are you using?

**Comment by sahmadreza** [comment]
> I'm sorry to see this. Honestly, this should be _impossible_ because `animatedBuilder` is called _after_ `initAnimation` which initializes `_fadeIn`.
> 
> **I am unable to reproduce this problem.** Are you able to reproduce this with the `example` application? It looks like you are using Windows for an Android target? What Android device target are you using?
> 
> @kw2019ltd Are you using Windows for an Android target too? What Android device target are you using?
Yes, I use windows 10

Device tested:

- Samsung c9000 android Api 26
- Android virtual device with Api 22 (Android studio AVD)

**Comment by aagarwal1012** [comment]
@sahmadreza, did you try `flutter clean` and rebuilding the app?

**Comment by sahmadreza** [comment]
> @sahmadreza, did you try `flutter clean` and rebuilding the app?


Yes , I tried again and the error is still there

```
The relevant error-causing widget was: 
  AnimatedBuilder file:///C:/Users/AHMADREZA-PC/AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/animated_text_kit-3.0.0/lib/src/animated_text.dart:183:13
When the exception was thrown, this was the stack: 
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      FadeAnimatedText.animatedBuilder (package:animated_text_kit/src/fade.dart:45:24)
#2      _AnimatedTextKitState.build.<anonymous closure> (package:animated_text_kit/src/animated_text.dart:187:32)
#3      AnimatedBuilder.build (package:flutter/src/widgets/transitions.dart:1443:19)
#4      _AnimatedState.build (package:flutter/src/widgets/transitions.dart:181:48)
...
════════════════════════════════════════════════════════════════════════════════════════════════════

```

Is there a problem with the dart version?

**Comment by awhitford** [comment]
@sahmadreza Can you please confirm:
* Does this issue appear with the `example` app?
* Does this issue appear while testing on a physical device?

I have a hypothesis that this is related to this line:
https://github.com/aagarwal1012/Animated-Text-Kit/blob/ef52c3f7f3447d88cc29706e5c8dbc06ac499c4c/lib/src/animated_text.dart#L220

However, I am not experiencing an issue with Android or the iOS Simulator on my Mac.  If this issue is limited to Windows, I wonder if this is really a Flutter bug because Flutter should be consistent across platforms.

**Comment by awhitford** [comment]
I submitted PR #169 that refactors the initialization sequence.  I highly suspect that the PR addresses this issue.  If it does not, then I suspect a Flutter bug with Windows.

**Comment by awhitford** [comment]
Version **3.0.1** has been released:  https://pub.dev/packages/animated_text_kit/changelog 🎉 
Please confirm if this resolves your problem with Windows.

**Comment by alittlebitofit** [comment]
Hi, I get the same error when I use `TypewriterAnimatedText`.
I am using:
* Ubuntu 20.10 LTS
* Samsung Galaxy M31, API level 29.
![animated_text_kit_AS](https://user-images.githubusercontent.com/42311431/101977226-fee00f80-3c71-11eb-9647-3fe0dce59675.png)

**Comment by kw2019ltd** [comment]
> 
> 
> Version **3.0.1** has been released: https://pub.dev/packages/animated_text_kit/changelog 🎉
> Please confirm if this resolves your problem with Windows.

working fine now no issue thanks

**Comment by awhitford** [comment]
@alittlebitofit Thanks for the feedback.  So this impacts Windows _AND_ Linux...  Can you please confirm if you can recreate this issue with the `example` app?

**Comment by awhitford** [comment]
@kw2019ltd OK, excellent, good news! 🎉

**Comment by AlienKevin** [comment]
I got the same error with `TypewriterAnimatedTextKit` using version **3.0.1**.

My platform:
* MacOS Big Sur
* iOS Simulator iPhone 12 Pro

Error message:
```
======== Exception caught by widgets library =======================================================
The following NoSuchMethodError was thrown building AnimatedBuilder(animation: AnimationController#d6c2c(▶ 0.619), dirty, state: _AnimatedState#2afa6):
The getter 'value' was called on null.
Receiver: null
Tried calling: value

The relevant error-causing widget was: 
  TypewriterAnimatedTextKit
```

**Comment by awhitford** [comment]
@AlienKevin Are you able to reproduce this issue with the `example` app?

I am unable to reproduce this issue.  I maintain that it _should be impossible_ because `build` should not be called before `initState`, and so the `initAnimation` will happen before `animatedBuild`.  However, I believe folks are running into this -- I just can't explain it. 😕

**Comment by AnirudhKK** [comment]
Having the same issue 

`TypewriterAnimatedTextKit(
                  text: ['Animated Text'],
                  textStyle: TextStyle(
                    fontSize: 45.0,
                    fontWeight: FontWeight.w900,
                  ),
                ),` 
gives 
`======== Exception caught by widgets library =======================================================
The getter 'value' was called on null.
Receiver: null
Tried calling: value
The relevant error-causing widget was: 
  TypewriterAnimatedTextKit file:///C:/Users/iaman/AndroidStudioProjects/flash-chat-flutter/lib/screens/welcome_screen.dart:54:17
====================================================================================================`

**Comment by awhitford** [comment]
@AnirudhKK Can you share your environment?  (OS, Flutter channel, Device Target?) Also, are you able to reproduce this with the `example` app?

**Comment by AnirudhKK** [comment]
Flutter 1.22.4 • channel stable • Android • Huawei JKM LX1 . Will let you know if I can reproduce it on example app.

**Comment by awhitford** [comment]
@AnirudhKK Are you using a Mac or Windows or Linux?

**Comment by AnirudhKK** [comment]
Windows 10 Pro 20H2

**Comment by AnirudhKK** [comment]
@awhitford again got the same error on Flutter Counter App

`The getter 'value' was called on null.
Receiver: null
Tried calling: value`

https://gist.github.com/AnirudhKK/3d932bce33ee7e672907068d0e7ba23d

**Comment by awhitford** [comment]
@AnirudhKK I appreciate the feedback.  I used the same code from your gist, but am **still unable to reproduce** or even logically explain how this could be possible.

Since you can recreate it...  Can you try adding a breakpoint on lines 40 and 65:
https://github.com/aagarwal1012/Animated-Text-Kit/blob/1a1f7ad39dd842509dd2d6d49380df7c915dc4dd/lib/src/typewriter.dart#L40

https://github.com/aagarwal1012/Animated-Text-Kit/blob/1a1f7ad39dd842509dd2d6d49380df7c915dc4dd/lib/src/typewriter.dart#L65

Then, are you seeing the breakpoint on line 65 stop _before_ line 40?  (Because that shouldn't be the case.  Line 40 should execute before line 65, so the `_typewriterText` variable will have been set, and there should be no issue.)

Line 40 should be called _once_ when the widget is added to the tree.  Then, line 65 is called frequently during `build`.

**Comment by ThinkDigitalSoftware** [comment]
I'm getting this error as well with  `animated_text_kit: ^3.0.1`. I only pasted your example in a Center in a scaffold.

```dart
// @dart=2.9
import 'package:animated_text_kit/animated_text_kit.dart';
import 'package:flutter/material.dart';

class Home extends StatefulWidget {
  Home({Key key}) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Center(
        child: SizedBox(
          width: 250.0,
          child: TypewriterAnimatedTextKit(
            onTap: () {
              print("Tap Event");
            },
            text: [
              "Discipline is the best tool",
              "Design first, then code",
              "Do not patch bugs out, rewrite them",
              "Do not test bugs out, design them out",
            ],
            textStyle: TextStyle(fontSize: 30.0, fontFamily: "Agne"),
            textAlign: TextAlign.start,
          ),
        ),
      ),
    );
  }
}
```

Flutter 1.25.0-9.0.pre.42 • channel master • https://github.com/flutter/flutter.git
Framework • revision a76bb1a08e (22 hours ago) • 2020-12-12 00:17:51 -0700
Engine • revision 4797b06652
Tools • Dart 2.12.0 (build 2.12.0-141.0.dev)

Running on macOS as a target device

**Comment by ThinkDigitalSoftware** [comment]
@awhitford,  Adding a `UniqueKey()` to my scaffold completely eliminated the problem. As I was hot restarting, each time, more and more errors would print out, taking longer and longer for the animation to start. Now this isn't an issue

**Comment by kw2019ltd** [comment]
hi sorry still getting same issue

I/flutter (11202): The getter 'value' was called on null.
I/flutter (11202): Receiver: null
I/flutter (11202): Tried calling: value
I/flutter (11202): 
I/flutter (11202): The relevant error-causing widget was:
I/flutter (11202):   FadeAnimatedTextKit-[<'13-12-20'>]
I/flutter (11202):   file:///C:/Users/kotai/AndroidStudioProjects/plants_price_flutter_app/lib/widget/plants/header/sticky_header_details.dart:280:18
I/flutter (11202): 
I/flutter (11202): When the exception was thrown, this was the stack:
I/flutter (11202): #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
I/flutter (11202): #1      FadeAnimatedText.animatedBuilder (package:animated_text_kit/src/fade.dart:45:24)
I/flutter (11202): #2      _AnimatedTextKitState.build.<anonymous closure> (package:animated_text_kit/src/animated_text.dart:186:32)
I/flutter (11202): #3      AnimatedBuilder.build (package:flutter/src/widgets/transitions.dart:1443:19)
I/flutter (11202): #4      _AnimatedState.build (package:flutter/src/widgets/transitions.dart:181:48)

**Comment by ThinkDigitalSoftware** [comment]
@kw2019ltd  did you hot **restart** already?

**Comment by alittlebitofit** [comment]
> @alittlebitofit Thanks for the feedback. Can you please confirm if you can recreate this issue with the `example` app? If not, then please share some source code that reproduces the issue.

@awhitford Thanks for the response.

`example` app works without any side-effect at all.

Ok, I'll share the source code: https://pastebin.com/MaUcqn0B

And this is the error output:  https://pastebin.com/zMYwDfjb

**Comment by awhitford** [comment]
I was finally able to reproduce this issue last night after switching to the dev channel.  I can trigger this issue and clearly see that `animatedBuilder` is running _before_ `initAnimation`, which still does not make sense to me. 😖

**Comment by ThinkDigitalSoftware** [comment]
Another issue I noticed is that if you did a set state while an animation is running and all the words in the animation receive null values for their fade in, fade out, duration, etc.

**Comment by awhitford** [comment]
I have distilled the issue down to a leaner reproducible test case, but still cannot explain what is going on.  I may have discovered a Flutter bug, but odds are that I am just not seeing the issue.  I ended up posting the issue on Stack Overflow for feedback: https://stackoverflow.com/questions/65285189/flutter-animatedbuilder-builder-bug

**Comment by AlienKevin** [comment]
## Example app
All animations work seamlessly
## My app
NoSuchMethodError
### Code
P.S. I tried removing my own `AnimationController` but it doesn't help.
```dart
class WelcomeScreen extends StatefulWidget {
  static String id = 'welcome';

  @override
  _WelcomeScreenState createState() => _WelcomeScreenState();
}

class _WelcomeScreenState extends State<WelcomeScreen>
    with SingleTickerProviderStateMixin {
  AnimationController controller;
  Animation<Color> animation;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(
      duration: Duration(milliseconds: 500),
      vsync: this,
    );

    animation = ColorTween(begin: Colors.blueGrey, end: Colors.white)
        .animate(controller);

    controller.addListener(() {
      setState(() {});
    });
    controller.forward();
  }

  @override
  void dispose() {
    super.dispose();
    controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: animation.value,
      body: Padding(
        padding: EdgeInsets.symmetric(horizontal: 24.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            Row(children: <Widget>[
              TypewriterAnimatedTextKit(
                text: ['Flash Chat'],
                textStyle: TextStyle(
                  fontSize: 45.0,
                  fontWeight: FontWeight.w900,
                ),
              ),
            ]),
          ],
        ),
      ),
    );
  }
}
```

**Comment by awhitford** [comment]
**Update**:  I now understand what is happening. 🎉 
Unfortunately, it is actually a non-trivial design flaw.  So now I need to refactor to fix it properly, and I'm experimenting with a few options.

**Comment by ThinkDigitalSoftware** [comment]
> **Update**: I now understand what is happening. 🎉
> Unfortunately, it is actually a non-trivial design flaw. So now I need to refactor to fix it properly, and I'm experimenting with a few options.

What's the basic issue? Maybe I know a good workaround to save some work?

**Comment by awhitford** [comment]
Thank you @ThinkDigitalSoftware for the offer.  My biggest issue was that I was unable to reproduce this, but I knew it was _real_ because so many complained.  I now can reliably recreate the bug.  In fact, it's so easy now that I'm astonished I didn't expose this issue before.  (Testing definitely needs to be tweaked.)

See the [SO article](https://stackoverflow.com/questions/65285189/flutter-animatedbuilder-builder-bug).  Basically, `setState` triggers a widget rebuild, which is messing with the `AnimatedText` class state.  In a nutshell, I have discovered (the hard way) _why_ StatefulWidgets are TWO classes.  `AnimatedText` is basically being rebuilt like a StatelessWidget, and then losing its state.  I erroneously thought that state would be managed by `AnimatedTextKitState`, but obviously that's not the case.  I do _finally_ understand what's going on.

I built a fix that works.  However, as I was transforming each implementation, it occurred to me whether I should simply make `AnimatedText` a `StatefulWidget` -- that way, I feel that I'm not recreating the wheel so much.  But the snag is how to best _hook_ into the _Animation Controller_ to add the `_animationEndCallback`.  The problem is that the `_animationEndCallback` is from the `AnimatedTextKit` parent, not the `AnimatedText` subclass.  I'm also debating on who should own the Animation Controller -- should each `AnimatedText` just create one, or should they use the `AnimatedTextKit`'s parent?  The devil is in the details...

Part of me would like to experiment with [flutter_hooks](https://pub.dev/packages/flutter_hooks), but the other part of me would like to avoid adding a dependency.

An annoying thing about this solution is that I ultimately need to tweak each `AnimatedText` subclass.  As a result, I want the _best_ solution.  I'm sad that some of the elegance of the v3 refactoring is going away, but I am retaining most of the key benefits (removed duplicate code, more flexible animation combinations, backwards compatibility).  And, I am seeing the possibility of folding `TextLiquidFill` to be like other `AnimatedText` subclasses.

I hope to resolve this today.

**Comment by ThinkDigitalSoftware** [comment]
You should look into using keys to maintain state. Adding them in the right place helps widgets identify their states and keeps them from rebuilding from scratch

**Comment by awhitford** [comment]
I did some experimentation with adding Keys in the beginning, but now that I understand the problem, it won't solve this particular problem.  I need to segregate the State so that it gets maintained.

**Comment by awhitford** [comment]
**Update**:  PR #176 fixes this issue. 🎉 🎉 🎉 
In the end, I figured out a _super simple solution_ that did not require modifying `AnimatedText` or any of its subclasses.

I updated the `example` app to show a _Tap Count_, which was a simple example of calling `setState` during an animation that triggers the widget rebuild and exposes this bug.  Before the `AnimatedTextKit` fix, I am able to reproduce the error.  After the fix, I am no longer able to reproduce the error.  This gives me confidence that this has been resolved.

I appreciate everyone's feedback, help, and patience.

@aagarwal1012 Let's work on releasing 3.0.2.
